src/VisualStudio/Core/Def/Implementation/Diagnostics/DiagnosticProgressReporter.cs

```diff
@@ -21,180 +21,180 @@ internal sealed class TaskCenterSolutionAnalysisProgressReporter
+        /// <summary>
+        /// Retrieve and throttle solution crawler events to be sent to the progress reporter UI.
+        /// 
+        /// there is no concurrent call to this method since ISolutionCrawlerProgressReporter will serialize all
+        /// events to preserve event ordering
+        /// </summary>
+        /// <param name="progressData"></param>
+        public void OnSolutionCrawlerProgressChanged(object sender, ProgressData progressData)
+        {
+            lock (_lock)
+            {
+                _lastProgressData = progressData;
+
+                // The task is running which will update the progress.
+                if (_intervalTask != null)
+                {
+                    return;
+                }
+
+                // Report progress immediately to ensure we update the UI on the first event.
+                ReportProgress();
+
+                // Kick off task to update the UI after a delay to pick up any new events.
+                _intervalTask = Task.CompletedTask.ContinueWithAfterDelay(() =>
+                {
+                    ReportProgress();
+                }, CancellationToken.None, s_minimumInterval, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
+            }
+        }
+
+        private void ReportProgress()
+        {
+            ProgressData data;
+            lock (_lock)
+            {
+                data = _lastProgressData;
+                _intervalTask = null;
+            }
+
+            UpdateUI(data);
+        }
+
+        private void UpdateUI(ProgressData progressData)
         {
-            // there is no concurrent call to this method since ISolutionCrawlerProgressReporter will serialize all
-            // events to preserve event ordering
             switch (progressData.Status)
             {
                 case ProgressStatus.Started:
-                    StartedOrStopped(started: true);
+                    Started();
                     break;
                 case ProgressStatus.PendingItemCountUpdated:
-                    _lastPendingItemCount = progressData.PendingItemCount.Value;
-                    ProgressUpdated();
+                    _lastProgressCount = progressData.PendingItemCount ?? 0;
+                    ChangeProgress(GetMessage(progressData, _lastProgressCount));
                     break;
                 case ProgressStatus.Stopped:
-                    StartedOrStopped(started: false);
+                    Stopped();
                     break;
                 case ProgressStatus.Evaluating:
                 case ProgressStatus.Paused:
-                    _lastProgressStatus = progressData.Status;
-                    ProgressUpdated();
+                    ChangeProgress(GetMessage(progressData, _lastProgressCount));
                     break;
                 default:
                     throw ExceptionUtilities.UnexpectedValue(progressData.Status);
             }
         }
```

Comment:
i'm skeptical about the need for this switch.  but will continue reviewing as if its necessary.